src/Tasks/AssemblyDependency/Node/RarTaskParameters.cs

```diff
@@ -0,0 +1,178 @@
+            {
+                object? value = property.Target.GetValue(rar);
+
+                if (IsParameterExplicitlySet(value, property.DefaultValue))
+                {
+                    taskParameters[property.Target.Name] = new TaskParameter(value);
+                }
+            }
+
+            return taskParameters;
+
+            // Determines if a parameter is explicitly set by comparing its current value against our cached default value.
+            static bool IsParameterExplicitlySet(object? value, object? defaultValue)
+            {
+                // .NET Framework doesn't have an allocation-free SequenceEqual, so manually compare array elements.
+                if (value is string[] stringArray
+                    && defaultValue is string[] defaultStringArray
+                    && stringArray.Length == defaultStringArray.Length)
+                {
+                    for (int i = 0; i < stringArray.Length; i++)
+                    {
+                        if (!string.Equals(stringArray[i], defaultStringArray[i], StringComparison.Ordinal))
+                        {
+                            return true;
+                        }
+                    }
+
+                    return false;
+                }
+
+                // Use Equals instead of '==' to correctly handle value equality.
+                return value?.Equals(defaultValue) == false;
+            }
+        }
+
+        /// <summary>
+        /// Sets each property on the RAR task from a previously extracted mapping of parameters.
+        /// </summary>
+        /// <param name="parameterType">The type of parameters in the dictionary.</param>
+        /// <param name="rar">The RAR task instance to set parameters on.</param>
+        /// <param name="parameters">A previously created mapping from parameter name to its wrapped value.</param>
+        internal static void Set(ParameterType parameterType, ResolveAssemblyReference rar, Dictionary<string, TaskParameter> parameters)
+        {
+            s_reflectedProperties ??= new ReflectedProperties();
+            ReflectedPropertyInfo[] properties = parameterType == ParameterType.Input
+                ? s_reflectedProperties.Inputs
+                : s_reflectedProperties.Outputs;
+
+            foreach (ReflectedPropertyInfo property in properties)
+            {
+                if (parameters.TryGetValue(property.Target.Name, out TaskParameter? parameter))
+                {
+                    property.Target.SetValue(rar, parameter.WrappedParameter);
+                }
+            }
+        }
+
+        private readonly struct ReflectedPropertyInfo(PropertyInfo target, object? defaultValue)
+        {
+            internal PropertyInfo Target { get; } = target;
+
+            internal object? DefaultValue { get; } = defaultValue;
+        }
+
+        private class ReflectedProperties
+        {
+            internal ReflectedProperties()
```

Comment:
Can you give some "why" background in a comment here--is this reflection-based for ease of modification, or does it have to be for some other reason, or?

Reply:
Originally for ease of modification, e.g. RAR getting a new input/output property and having things automatically work. Although tbh I'm not as attached to the idea since there's a fair amount of special casing that needed to happen either way.